src/Type/Php/IsNumericFunctionTypeSpecifyingExtension.php

```diff
@@ -49,7 +50,12 @@ public function specifyTypes(FunctionReflection $functionReflection, FuncCall $n
 			$numericTypes[] = new StringType();
 		}
 
-		return $this->typeSpecifier->create($node->args[0]->value, new UnionType($numericTypes), $context);
+		$numericTypes = new IntersectionType([
+			new UnionType($numericTypes),
+			$node->args[0]->getType(),
+		]);
+
+		return $this->typeSpecifier->create($node->args[0]->value, $numericTypes, $context);
```

Comment:
Sorry, but I don't see how this could work. There's several mistakes:

1) You don't get PHPStan's type by calling `Arg_::getType()` - you need to get `$scope->getType($node->args[0]->value)` but in this case it wouldn't do anything because:
2) This is actually done for all TypeSpecifier outputs - types from type-specifying extensions are intersected by default somewhere in `Scope::filterBySpecifiedTypes`.
3) If you're instantiating `IntersectionType` directly instead of `TypeCombinator::intersect()`, you're risking un-normalized types.

So this change is basically no-op, I really don't see how it helps fixing #3133.

The problem that's here is that in case of https://phpstan.org/r/d65e6a61-95bd-44ac-8093-5afe354bf16c the code probably ends in `return new SpecifiedTypes([], []);` on line 41. That's what's wrong with the extension.

Reply:
Thanks for the fast and detailed reply. Where should I add a regression test with your snippet to start fixing it?

Reply:
The best way is to look at NodeScopeResolver::testFileAsserts and add another data provider :) you should `assertType` after the `!is_numeric` condition

Reply:
Done, but I can't make it work, test is broken in master right now. I did it on top of the last stable tag here: https://github.com/phpstan/phpstan-src/commit/ad96bb9df257953da4bb30887d91d0fca369449d

But from that, I can't go further, I don't understand a few pieces. What's the `$context` here? Do I have restrict the `SpecifiedTypes` to exclude the non numeric types? How?

Reply:
Tests on master aren't broken AFAIK: https://github.com/phpstan/phpstan-src/actions/runs/78827288

If I knew how exactly to fix this, I'd already done it :) For starters, try to revert the extension a few changes back (https://github.com/phpstan/phpstan-src/commits/1a1f5d6f0e93caaf8fcc7fea946efd62e1d73456/src/Type/Php/IsNumericFunctionTypeSpecifyingExtension.php), I bet it will fix your regression test but make some other tests fail.